bundles/admin/admin-hierarchical-layerlist/instance.js

```diff
@@ -0,0 +1,364 @@
+                var group = me.layerService.getAllLayerGroups(groupId);
+                var options = {
+                    locale: group.name,
+                    selectable: group.selectable
+                };
+                options.type = 'subgroup';
+
+                var popupConf = me.group.getGroupAddingPopupConf(tool, groupId, parentId, options);
+                var popup = popupConf.popup;
+                var message = popupConf.message;
+                popupConf.popup.show(me.locale('groupTitles.editSubgroup'), message, popupConf.buttons);
+                popupConf.popup.makeModal();
+            }, {
+                cls: 'edit-subgroup',
+                tooltip: me.locale('tooltips.editSubgroup')
+            });
+
+            // Add new tool to adding sub-groups
+            me.service.addSubgroupTool('add-subgroup-subgroup', function(tool, parentId) {
+                var popupConf = me.group.getGroupAddingPopupConf(tool, null, parentId, {
+                    type: 'subgroup-subgroup'
+                });
+
+                var popup = popupConf.popup;
+                var message = popupConf.message;
+                popupConf.popup.show(me.locale('groupTitles.addSubgroup'), message, popupConf.buttons);
+                popupConf.popup.makeModal();
+            }, {
+                cls: 'add-subgroup-subgroup',
+                tooltip: me.locale('tooltips.addSubgroup')
+            });
+        },
+
+        /**
+         * Add subgroup  subgroup tools
+         * @method  _addSubgroupSubgroupTools
+         * @private
+         */
+        _addSubgroupSubgroupTools: function() {
+            var me = this;
+            // Add layer add tools
+            me.service.addSubgroupSubgroupTool('add-layer', function(tool, groupId, parentId) {
+                me.layer.showLayerAddPopup(tool, null, groupId);
+            }, {
+                cls: 'add-layer',
+                tooltip: me.locale('tooltips.addLayer')
+            });
+
+            // Add edit tool to adding groups
+            me.service.addSubgroupSubgroupTool('edit-subgroup-subgroup', function(tool, groupId, parentId) {
+                var group = me.layerService.getAllLayerGroups(groupId);
+                var options = {
+                    locale: group.name,
+                    selectable: group.selectable
+                };
+                options.type = 'subgroup-subgroup';
+
+                var popupConf = me.group.getGroupAddingPopupConf(tool, groupId, parentId, options);
+                var popup = popupConf.popup;
+                var message = popupConf.message;
+                popupConf.popup.show(me.locale('groupTitles.editSubgroup'), message, popupConf.buttons);
+                popupConf.popup.makeModal();
+            }, {
+                cls: 'edit-subgroup-subgroup',
+                tooltip: me.locale('tooltips.editSubgroup')
+            });
```

Comment:
You don't need to change anything here, but did I get this right?

- On the main level you can add a group
- On group level you can add a layer, edit the group and add a subgroup
- On subgroup level you can add a layer, edit the subgroup and add a subgroup subgroup
- On subgroup subgroup level you can add a layer, edit the subgroup subgroup

Does adding a group differ from adding a subgroup or subgroup subgroup? Seems this could have been written without hardcoded levels:

- on main level you can only add a group
- each group can add a layer, edit the group and add a subgroup (except for adding subgroup is not allowed on configurable max depth for group)

Reply:
Yes these are right.  In hierarchical-layerside side there is container for tools. Jstree has also typed nodes. Main level can contains also other tools, for example need i add update all getcapabilities here ?